components/normalizer/src/lib.rs

```diff
@@ -1608,5 +1621,262 @@ impl ComposingNormalizer {
+pub struct CanonicalComposition {
+    canonical_compositions: DataPayload<CanonicalCompositionsV1Marker>,
+}
+
+impl CanonicalComposition {
+    /// Performs canonical composition (including Hangul) on a pair of
+    /// characters or returns `None` if these characters don't compose.
+    /// Composition exclusions are taken into account.
+    #[inline]
+    pub fn compose(&self, starter: char, second: char) -> Option<char> {
+        let v = u32::from(second).wrapping_sub(HANGUL_V_BASE);
+        if v >= HANGUL_T_LIMIT - HANGUL_V_BASE {
+            return self.compose_non_hangul(starter, second);
+        }
+        if v < HANGUL_V_COUNT {
+            let l = u32::from(starter).wrapping_sub(HANGUL_L_BASE);
+            if l < HANGUL_L_COUNT {
+                let lv = l * HANGUL_N_COUNT + v * HANGUL_T_COUNT;
+                // Safe, because the inputs are known to be in range.
+                return Some(unsafe { char::from_u32_unchecked(HANGUL_S_BASE + lv) });
+            }
+            return None;
+        }
+        if in_inclusive_range(second, '\u{11A8}', '\u{11C2}') {
+            let lv = u32::from(starter).wrapping_sub(HANGUL_S_BASE);
+            if lv < HANGUL_S_COUNT && lv % HANGUL_T_COUNT == 0 {
+                let lvt = lv + (u32::from(second) - HANGUL_T_BASE);
+                // Safe, because the inputs are known to be in range.
+                return Some(unsafe { char::from_u32_unchecked(HANGUL_S_BASE + lvt) });
+            }
+        }
+        None
+    }
+
+    /// Performs non-Hangul canonical composition on a pair of characters
+    /// or returns `None` if these characters don't compose. Composition
+    /// exclusions are taken into account. (The return value is `None` with
+    /// Hangul arguments.)
+    #[inline(always)]
+    fn compose_non_hangul(&self, starter: char, second: char) -> Option<char> {
+        compose_non_hangul(
+            self.canonical_compositions
+                .get()
+                .canonical_compositions
+                .iter(),
+            starter,
+            second,
+        )
+    }
+
+    /// Construct from data provider.
+    pub fn try_new<D>(data_provider: &D) -> Result<Self, NormalizerError>
+    where
+        D: ResourceProvider<CanonicalCompositionsV1Marker> + ?Sized,
+    {
+        let canonical_compositions: DataPayload<CanonicalCompositionsV1Marker> = data_provider
+            .load_resource(&DataRequest::default())?
+            .take_payload()?;
+        Ok(CanonicalComposition {
+            canonical_compositions,
+        })
+    }
+}
+
+/// The outcome of non-recursive canonical decomposition of a character.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Decomposed {
+    /// The character is its own canonical decomposition.
+    Default,
+    /// The character decomposes to a single different character.
+    Singleton(char),
+    /// The character decomposes to two characters.
+    Expansion(char, char),
```

Comment:
Question: is the decomposition defined by Unicode to be a max of only 2 code points?

Reply:
Yes. [UAX 44, 5.7.3](https://www.unicode.org/reports/tr44/#Character_Decomposition_Mappings) (emphasis mine):
> In some instances a canonical mapping or a compatibility mapping may consist of a single character. For a canonical mapping, this indicates that the character is a canonical equivalent of another single character. For a compatibility mapping, this indicates that the character is a compatibility equivalent of another single character.
> 
> _A canonical mapping may also consist of a pair of characters, but is never longer than two characters._ When a canonical mapping consists of a pair of characters, the first character may itself be a character with a decomposition mapping, but the second character never has a decomposition mapping.